src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -438,11 +483,198 @@ StoragePtr DatabaseCatalog::tryGetTable(const StorageID & table_id) const
+            create->database = table_id.database_name;
+            create->table = table_id.table_name;
+            try
+            {
+                table = createTableFromAST(*create, table_id.getDatabaseName(), data_path, *global_context, false).second;
+            }
+            catch (...)
+            {
+                tryLogCurrentException(log, "Cannot load partially dropped table " + table_id.getNameForLogs() +
+                                            " from: " + dropped_metadata_path +
+                                            ". Parsed query: " + serializeAST(*create) +
+                                            ". Will remove metadata and " + data_path +
+                                            ". Garbage may be left in ZooKeeper.");
+            }
+        }
+        else
+        {
+            LOG_WARNING(log, "Cannot parse metadata of partially dropped table "
+                    << table_id.getNameForLogs() << " from " << dropped_metadata_path
+                    << ". Will remove metadata file and data directory. Garbage may be left in /store directory and ZooKeeper.");
+        }
+
+        drop_time = Poco::File(dropped_metadata_path).getLastModified().epochTime();
+    }
+
+    std::lock_guard lock(tables_marked_dropped_mutex);
+    if (ignore_delay)
+        tables_marked_dropped.push_front({table_id, table, dropped_metadata_path, 0});
+    else
+        tables_marked_dropped.push_back({table_id, table, dropped_metadata_path, drop_time});
+}
+
+void DatabaseCatalog::dropTableDataTask()
+{
+    TableMarkedAsDropped table;
+    try
+    {
+        std::lock_guard lock(tables_marked_dropped_mutex);
+        time_t current_time = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+        auto it = std::find_if(tables_marked_dropped.begin(), tables_marked_dropped.end(), [&](const auto & elem)
+        {
+            bool not_in_use = !elem.table || elem.table.unique();
+            bool old_enough = elem.drop_time + drop_delay_s < current_time;
+            return not_in_use && old_enough;
+        });
+        if (it != tables_marked_dropped.end())
+        {
+            table = std::move(*it);
+            LOG_INFO(log, "Will try drop " + table.table_id.getNameForLogs());
+            tables_marked_dropped.erase(it);
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, __PRETTY_FUNCTION__);
+    }
+
+    if (table.table_id)
+    {
+
+        try
+        {
+            dropTableFinally(table);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Cannot drop table " + table.table_id.getNameForLogs() +
+                                        ". Will retry later.");
+            {
+                std::lock_guard lock(tables_marked_dropped_mutex);
+                tables_marked_dropped.emplace_back(std::move(table));
+            }
+        }
+    }
+
+    (*drop_task)->scheduleAfter(reschedule_time_ms);
```

Comment:
Maybe we should do it only when we have `tables_marked_dropped` not empty? And also schedule it in each `drop` call. ?